### Creating the initial dataset

The main objective here is to explore and merge datasets to create a final dataframe containing:

* sale_date: when the product was sold
* sku_id: the unique identifier of the product
* seller_id: who sold the product
* qty: how much was sold (quantity in units)

I'll use this simple dataset to perform the further steps of my project. It is important to state that, for this project, I'll only need the temporal features that can be created using the columns cited above.The dataset documentation from kaggle has this image: 

![](https://i.imgur.com/HRhd2Y0.png)

In [35]:
## importing libs

import pandas as pd

In [36]:
## reading the order_items dataset

df = pd.read_csv('../data/raw/olist_order_items_dataset.csv')
df.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [37]:
## inspecting the dataset shape

df.shape

(112650, 7)

In [38]:
## describing all columns from the dataset

df.describe(include='all')

order_id  order_item_id  \
count                             112650  112650.000000   
unique                             98666            NaN   
top     8272b63d03f5f79c56e9e4120aec44ef            NaN   
freq                                  21            NaN   
mean                                 NaN       1.197834   
std                                  NaN       0.705124   
min                                  NaN       1.000000   
25%                                  NaN       1.000000   
50%                                  NaN       1.000000   
75%                                  NaN       1.000000   
max                                  NaN      21.000000   

                              product_id                         seller_id  \
count                             112650                            112650   
unique                             32951                              3095   
top     aca2eb7d00ea1a7b8ebd4e68314663af  6560211a19b47992c3666cc44a7e94c0   
freq                                 527                              2033   
mean                                 NaN                               NaN   
std                                  NaN                               NaN   
min                                  NaN                               NaN   
25%                                  NaN                               NaN   
50%                                  NaN                               NaN   
75%                                  NaN                               NaN   
max                                  NaN                               NaN   

        shipping_limit_date          price  freight_value  
count                112650  112650.000000  112650.000000  
unique                93318            NaN            NaN  
top     2017-07-21 18:25:23            NaN            NaN  
freq                     21            NaN            NaN  
mean                    NaN     120.653739      19.990320  
std                     NaN     183.633928      15.806405  
min                     NaN       0.850000       0.000000  
25%                     NaN      39.900000      13.080000  
50%                     NaN      74.990000      16.260000  
75%                     NaN     134.900000      21.150000  
max                     NaN    6735.000000     409.680000

The order_id column has 112650 entries, but only 98666 (87% of the total) entries are unique. This difference can be attributed by the fact a order with multiple itens generate multiple lines (n lines -> n products within the same order).

In [39]:
df[df.seller_id == '4a3ca9315b744ce9f8e9374361493884'].order_id.value_counts()

b8602ecb5d10f27cd89d0e7b5a8ec2a8    5
26a14aeb2410adbbf74a72c4ef0c8ae3    4
c887c09a3acf3b6fe0ba319d68d385a8    4
16febac56c216c1cc40407ff81315a97    4
e1bc1083cd7acd30d0576335373b907d    4
                                   ..
580c882a5805c78546ea65ec6d3690e4    1
58076d990830fefdf892853e25db27c8    1
57bfb7a49bcdecb20d9dbb778d7d7590    1
57638520434d4c12b81d532ad9bf2f1f    1
fff2e9e3aa8644e19710216b4ef53ab2    1
Name: order_id, Length: 1806, dtype: int64

If my logic is correct, the order "b8602ecb5d10f27cd89d0e7b5a8ec2a8" will 5 itens.

In [40]:
df[df.order_id == "b8602ecb5d10f27cd89d0e7b5a8ec2a8"]

order_id  order_item_id  \
81098  b8602ecb5d10f27cd89d0e7b5a8ec2a8              1   
81099  b8602ecb5d10f27cd89d0e7b5a8ec2a8              2   
81100  b8602ecb5d10f27cd89d0e7b5a8ec2a8              3   
81101  b8602ecb5d10f27cd89d0e7b5a8ec2a8              4   
81102  b8602ecb5d10f27cd89d0e7b5a8ec2a8              5   

                             product_id                         seller_id  \
81098  f2e53dd1670f3c376518263b3f71424d  4a3ca9315b744ce9f8e9374361493884   
81099  f2e53dd1670f3c376518263b3f71424d  4a3ca9315b744ce9f8e9374361493884   
81100  f2e53dd1670f3c376518263b3f71424d  4a3ca9315b744ce9f8e9374361493884   
81101  f2e53dd1670f3c376518263b3f71424d  4a3ca9315b744ce9f8e9374361493884   
81102  f2e53dd1670f3c376518263b3f71424d  4a3ca9315b744ce9f8e9374361493884   

       shipping_limit_date  price  freight_value  
81098  2017-11-14 16:30:36  109.9          17.48  
81099  2017-11-14 16:30:36  109.9          17.48  
81100  2017-11-14 16:30:36  109.9          17.48  
81101  2017-11-14 16:30:36  109.9          17.48  
81102  2017-11-14 16:30:36  109.9          17.48

As we can see above the order indeed has 5 itens. This also shows that even if the products are the same and are bought in the same order they will generate multiple lines. Merging this order_item dataset with the order_dataset we can create the dataframe that is needed to continue the project.

In [41]:
final_df = (
    pd.read_csv('../data/raw/olist_orders_dataset.csv')
    .query("order_status == 'delivered'")
    .loc[:,['order_id','order_approved_at']]
    .merge(df.loc[:,['order_id','product_id','seller_id']], how='right')
    .assign(sale_date = lambda x: pd.to_datetime(x.order_approved_at).dt.date)
    .groupby(['sale_date','seller_id','product_id']).count()
    .reset_index()
    .drop("order_approved_at", axis=1)
    .rename(columns={'order_id':'qty', 'product_id':'sku_id'})
)

In [42]:
final_df.head()

sale_date                         seller_id  \
0  2016-09-15  ecccfa2bb93b34a3bf033cc5d1dcdc69   
1  2016-10-04  4b1eaadf791bdbbad8c4a35b65236d52   
2  2016-10-04  5b179e9e8cc7ab6fd113a46ca584da81   
3  2016-10-04  620c87c171fb2a6dd6e8bb4dec959fc6   
4  2016-10-04  624f4ece8da4aafb77699233d480f8ef   

                             sku_id  qty  
0  5a6b04657a4c5ee34285d1e4619a96b4    3  
1  33430c5c1027d812b5c62f778e5ee7f7    1  
2  29bedbfbe73ea473396306b4f0c010ec    1  
3  eba7488e1c67729f045ab43fac426f2e    1  
4  817e1c2d22418c36386406ccacfa53e8    1

In [43]:
final_df.isna().sum()

sale_date    0
seller_id    0
sku_id       0
qty          0
dtype: int64

In [44]:
final_df.to_feather('../data/interim/order_dataset_to_forecastability')